### DISCUSSION TITLE
Dynamic metadata, dedupe requests but keep data fresh

### DISCUSSION THREAD

**Discussion by generalentropy** [vercel/next.js#72472] [Help]
### Summary

How to set dynamic metadata without duplicating requests and keep my data up to date? 

I'm not able to do both.

I've tried different cache options: "no-store", "no-cache", and revalidate: 1.

I've also tried export const dynamic = 'force-dynamic' (which seems useless as I'm using dynamic params).

But the end result remains the same: duplicate requests (sometimes 3, depending on the configuration) or stale data.

_Server component_ :

```
const getProduct = async (id: string) => {
  const res = await fetch(`${BASE_URL}/api/products/unique?id=${id}`);
  return await res.json();
};

export async function generateMetadata({ params }: { params: { id: string } }) {
  const { id } = params;

  const productData: Product | null = await getProduct(id);

  return {
    title: `${productData?.name}`,
    description: productData?.description,
  };
}

async function ProductIdRoute({ params }: { params: { id: string } }) {
  const { id } = params;

  const productData: Product | null = await getProduct(id);

  if (!productData)
    return...
```

**Comment by jonas-is-coding** [comment]
Hey, I think your error occures because you use ```no-store``` and ```no-cache``` with dynamic data.

## Solution
```javascript
import { cache } from 'react';
import { headers } from 'next/headers';

// Cache the product request with the react caching function...
const getProduct = cache(async (id: string) => {
  const headersList = headers();
  const timestamp = headersList.get('x-timestamp') || Date.now();
  
  const res = await fetch(
    `${BASE_URL}/api/products/unique?id=${id}&_t=${timestamp}`,
    {
      next: {
       // Use nextjs revalidation...
        revalidate: 60, // 60 Sekunden
        tags: [`product-${id}`],
      }
    }
  );

  if (!res.ok) {
    throw new Error('Failed to fetch product');
  }

  return res.json();
});

export async function generateMetadata({ params }: { params: { id: string } }) {
  const { id } = params;
  
  try {
    const productData = await getProduct(id);
    
    if (!productData) {
      return {
        title: 'Product not found',
        description: 'Die requested product could not be found.'
      };
    }

    return {
      title: productData.name,
      description: productData.description,
    };
  } catch (error) {
    return {
      title: 'Error',
      description: 'An error ocurred'
    };
  }
}
```

## Explanation
1. React cache():
Prevents duplicate requests within a render execution
Automatically caches the results in memory
2. Next.js caching strategy:
revalidate: 60 enables incremental static regeneration
tags allow targeted invalidation of the cache
3. Error handling:
Robust error handling in both functions
Fallback metadata for errors
4. Optimisations:
Timestamp parameter for cache busting if necessary
Deduplication of fetch requests

**Comment by generalentropy** [comment]
Hi Jonas, thank you

With your exact implementation (including the timestamp), there are two API calls. without timestamp, one call, but the data becomes stale.

The problem seems to be the same, whether I'm using React cache() or Next native deduplication mechanism.

My main issue is serving fresh data _every time_ my route is visited. I've noticed that there's always cached data somewhere due to Next.js's caching policy, which causes a delay in displaying fresh data.

The most radical solution I found to ensure 100% fresh data on every visit is to add a client component within my server component that uses a useEffect to call a server action with revalidatePath or revalidateTag. This works effectively but seems extremely hacky.

Another option could be to use only a client component with React Query, but I would lose all the advantages of server components.
